solana-labs/wallet-adapter
# Add Vue core package

Hi there 👋 

This PR adds a new core package that supports Vue 3.

It works by providing two functions:
- `initWallet` that initialises the wallet store and should be called only once.
- `useWallet` that returns that store by reference so we can use its content anywhere in our application.

### Basic usage

```js
import { initWallet, useWallet } from '@solana/wallet-adapter-vue';
import { getPhantomWallet } from '@solana/wallet-adapter-wallets';

// Initialise the wallet store once.
const wallets = [getPhantomWallet()];
initWallet({ wallets, autoConnect: true });

// Use its content anywhere.
const { wallet, select, connect, disconnect, connected, sendTransaction } = useWallet();
select('Phantom');
connect();
sendTransaction(...);
```

### References

Analogously to the React core package, here are the options you can provide to the `initWallet` method.

```ts
interface WalletStoreProps {
    wallets: Wallet[];
    autoConnect?: boolean;                  // Default: false
    onError?: (error: WalletError) => void; // Default: (error: WalletError) => console.error(error)
    localStorageKey?: string;               // Default: "walletName"
}
```

And here is the definition of the wallet store you can access from anywhere in your Vue application once initialised.

```ts
interface WalletStore {
    // Props.
    wallets: Wallet[];
    autoConnect: boolean;

    // Data.
    walletName: Ref<string | null>,
    walletsByName: Ref<WalletDictionary>,
    wallet: Ref<Wallet | null>;
    adapter: Ref<Adapter | null>;
    publicKey: Ref<PublicKey | null>;
    ready: Ref<boolean>;
    connected: Ref<boolean>;
    connecting: Ref<boolean>;
    disconnecting: Ref<boolean>;

    // Methods.
    select(walletName: WalletName): void;
    connect(): Promise<void>;
    disconnect(): Promise<void>;
    sendTransaction(transaction: Transaction, connection: Connection, options?: SendTransactionOptions): Promise<TransactionSignature>;
    signTransaction(transaction: Transaction): Promise<Transaction>;
    signAllTransactions(transaction: Transaction[]): Promise<Transaction[]>;
    signMessage(message: Uint8Array): Promise<Uint8Array>;
}
```

### Why not provide/inject?
In React, it is common to use Context components and retrieve their content within their descendants. Whilst Vue has a similar concept via the `provide` and `inject` API, they aren't commonly used in the Vue community. Instead, it is more common to create a global store using the composition API and access its content from anywhere when needed, which is why I went for that approach in this PR.

### The `isolatedModules` flag
Note that I had to override the following TS option `"isolatedModules": false` inside the vue package in order to import from `@vue/runtime-core` (See https://github.com/solana-labs/wallet-adapter/issues/67#issuecomment-929386709).

### Manual testing
I manually tested this package locally [in my solana test app](https://github.com/lorisleiva/solana-vue3-counter/tree/feature/test-local-vue-adapter) by installing it as a symlink — i.e. using `"@solana/wallet-adapter-vue": "file:../../wallet-adapter/packages/core/vue"`. Everything worked properly and I was able to sign and send some transactions using my Phantom wallet.

However, I noticed that components using the references (`Ref`) provided by the store did not update when the references changed. I can see the references updating in real time in the Vue dev tools but for some reason the component itself does not re-render itself. When I call `this.$forceUpdate` on the component, I can see all the latest values being updated properly. I believe this is just a quirk from installing the package locally using a symlink but I thought I'd mention it in case this issue ring a bell to someone.
